src/transactions/CreateClaimableBalanceOpFrame.cpp

```diff
@@ -0,0 +1,274 @@
+            return false;
+        }
+
+        auto ok = addBalance(header, sourceAccount, -amount);
+        assert(ok);
+    }
+    else
+    {
+        auto trustline = loadTrustLine(ltx, getSourceID(), asset);
+        if (!trustline)
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_NO_TRUST);
+            return false;
+        }
+        if (!trustline.isAuthorized())
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_NOT_AUTHORIZED);
+            return false;
+        }
+
+        if (!trustline.addBalance(header, -amount))
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_UNDERFUNDED);
+            return false;
+        }
+    }
+
+    // Create claimable balance entry
+    LedgerEntry newClaimableBalance;
+    newClaimableBalance.data.type(CLAIMABLE_BALANCE);
+
+    auto& claimableBalanceEntry = newClaimableBalance.data.claimableBalance();
+    claimableBalanceEntry.createdBy = getSourceID();
+    claimableBalanceEntry.amount = amount;
+    claimableBalanceEntry.asset = asset;
+    claimableBalanceEntry.reserve = entryCost;
+
+    OperationID operationID;
+    operationID.type(ENVELOPE_TYPE_OP_ID);
+    operationID.id().sourceAccount = toMuxedAccount(getSourceID());
+    operationID.id().seqNum = mParentTx.getSeqNum();
+    operationID.id().opNum = mOpIndex;
+
+    claimableBalanceEntry.balanceID.v0() =
+        sha256(xdr::xdr_to_opaque(operationID));
+
+    claimableBalanceEntry.claimants = claimants;
+    for (auto& claimant : claimableBalanceEntry.claimants)
+    {
+        updatePredicatesForApply(claimant.v0().predicate,
+                                 header.current().scpValue.closeTime);
+    }
+
+    ltx.create(newClaimableBalance);
+
+    innerResult().code(CREATE_CLAIMABLE_BALANCE_SUCCESS);
+    return true;
+}
+
+bool
+CreateClaimableBalanceOpFrame::doCheckValid(uint32_t ledgerVersion)
+{
+    auto const& claimants = mCreateClaimableBalance.claimants;
+
+    if (!isAssetValid(mCreateClaimableBalance.asset) ||
+        mCreateClaimableBalance.amount <= 0 || claimants.empty())
+    {
+        innerResult().code(CREATE_CLAIMABLE_BALANCE_MALFORMED);
+        return false;
+    }
+
+    // check for duplicates
+    std::unordered_set<AccountID> dests;
+    for (auto const& claimant : claimants)
+    {
+        auto const& dest = claimant.v0().destination;
+        if (!dests.emplace(dest).second)
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_MALFORMED);
+            return false;
+        }
+    }
+
+    for (auto const& claimant : claimants)
+    {
+        if (!validatePredicate(claimant.v0().predicate, 1))
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_MALFORMED);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void
+CreateClaimableBalanceOpFrame::insertLedgerKeysToPrefetch(
+    std::unordered_set<LedgerKey>& keys) const
+{
+    // Prefetch issuer for non-native assets
```

Comment:
comment is wrong, you're prefetching the trust line

Reply:
fixed